Add Pages Early Hints docs (#5989)
cloudflare/cloudflare-docs: content/pages/platform/early-hints.md

Early Hints automatically caches any [`preload`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) and [`preconnect`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preconnect) type [`Link` headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link) to send as Early Hints to the browser. The hints are sent to the browser before the full response is prepared, and the browser can figure out how to load the webpage faster for the end user. There are two ways to create these `Link` headers in Pages:

## 1. Configure your `_headers` file

Create custom headers using the [`_headers` file](/pages/platform/headers/). For example, if you include a particular stylesheet on your `/blog/` section of your website, you can create the following rule:

```txt
---
filename: _headers
---
/blog/*
  Link: </styles.css>; rel=preload; as=stylesheet
```

> Pages will then attach this `Link: </styles.css>; rel=preload; as=stylesheet` header, which Early Hints will then emit as an Early Hint once cached.

Reviewer: ```suggestion
Pages will then attach this `Link: </styles.css>; rel=preload; as=stylesheet` header. Early Hints will then emit this header as an Early Hint once cached.
```

## 2. Automatic `Link` header generation

In order to make the authoring experience easier, Pages also automatically generates `Link` headers from any `<link>` HTML elements with any of the following attributes:

- `href`
- `as`
- `rel` (`preload` or `preconnect`)

> Note: `<link>` elements which contain any other additional attributes (e.g. `fetchpriority`, `crossorigin` or `data-do-not-generate-a-link-header`) will not be used to generate `Link` headers in order to prevent accidentally losing any custom prioritization logic that would otherwise be dropped as an Early Hint.

Reviewer: ```suggestion
`<link>` elements which contain any other additional attributes (for example, `fetchpriority`, `crossorigin` or `data-do-not-generate-a-link-header`) will not be used to generate `Link` headers in order to prevent accidentally losing any custom prioritization logic that would otherwise be dropped as an Early Hint.
```

Reviewer: If it needs to be a note, then we can make this an aside. Otherwise, taking out the "note".

> This allows you to directly create Early Hints as you're authoring your document, without needing to flick between your HTML and `_headers` file.

Reviewer: ```suggestion
This allows you to directly create Early Hints as you are writing your document, without needing to alternate between your HTML and `_headers` file.
```

```html
---
filename: /blog/index.html
---

<html>
  <head>
    <link rel="preload" href="/style.css" as="stylesheet" />
    <link rel="stylesheet" href="/style.css" />
  </head>
</html>
```

### Disable automatic `Link` header generation Automatic `Link` header

> Automatic `Link` header generation should not have any negative performance impact on your website. If you find yourself needing to disable this feature, please reach let us know about your circumstance in our [Discord server](https://discord.com/invite/cloudflaredev).

Reviewer: ```suggestion
Automatic `Link` header generation should not have any negative performance impact on your website. If you need to disable this feature, contact us by letting us know about your circumstance in our [Discord server](https://discord.com/invite/cloudflaredev).
```

---
pcx_content_type: concept
title: Early Hints
---

# Early Hints

> [Early Hints](https://developers.cloudflare.com/cache/about/early-hints/) is enabled automatically on all `pages.dev` domains. [It can be enabled for your any custom domains](https://developers.cloudflare.com/cache/about/early-hints/#enabling-early-hints) by navigating to **Speed** > **Optimization** and toggling the "Early Hints" setting in the dashboard.

Reviewer: ```suggestion
[Early Hints](https://developers.cloudflare.com/cache/about/early-hints/) help the browser to load webpages faster. Early Hints is enabled automatically on all `pages.dev` domains. It can be [enabled](https://developers.cloudflare.com/cache/about/early-hints/#enabling-early-hints) for your custom domains by logging into the [Cloudflare dashboard](https://dash.cloudflare.com) > selecting your account and domain > **Speed** > **Optimization** and turning on the **Early Hints** toggle.
```

> Early Hints automatically caches any [`preload`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) and [`preconnect`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preconnect) type [`Link` headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link) to send as Early Hints. There are two ways to create these `Link` headers in Pages:

Reviewer: ```suggestion
Early Hints automatically caches any [`preload`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) and [`preconnect`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preconnect) type [`Link` headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link) to send as Early Hints to the browser. The hints are sent to the browser before the full response is prepared, and the browser can figure out how to load the webpage faster for the end user. There are two ways to create these `Link` headers in Pages:
```

> 

Reviewer: There needs to be a sentence here that addresses what the user is about to go. "This guide will teach you _____" "By following this guide, you will _____" before the steps.

Author: This isn't a guide (the user doesn't follow step 1 and then step 2), rather, this presents two options. They can choose to do either step 1, or step 2, or both.

```suggestion

Early Hints can be created with either of the two methods detailed below.

```

Reviewer: I'd resolve my comment but it doesn't give me that option. This is good as is, doesn't need changes :)

> ## 1. Using a `_headers` file

Reviewer: ```suggestion
## 1. Configure your `_headers` file
```

Author: Does this still make sense if this isn't a mandatory step in a guide?

> Custom headers can be created by using the [`_headers` file](/pages/platform/headers/). For example, if you include a particular stylesheet on your `/blog/` section of your website, you can create the following rule:

Reviewer: ```suggestion
Create custom headers using the [`_headers` file](/pages/platform/headers/). For example, if you include a particular stylesheet on your `/blog/` section of your website, you can create the following rule:
```

Author: Same here.

```txt
---
filename: _headers
---
/blog/*
  Link: </styles.css>; rel=preload; as=stylesheet
```

Pages will then attach this `Link: </styles.css>; rel=preload; as=stylesheet` header, which Early Hints will then emit as an Early Hint once cached.

## 2. Automatic `Link` header generation

> In order to make the authoring experience easier, Pages also automatically generates `Link` headers from any `<link>` HTML elements with any of the following attributes:

Reviewer: Is the user doing anything in this step?

Author: No. This is one of the mechanisms that generates Early Hints. The only action a user can take here is below under the "Disabling automatic `Link` header generation Automatic `Link` header" header, where they can essentially turn off this automatic behavior if they choose.

- `href`
- `as`
- `rel` (`preload` or `preconnect`)

Note: `<link>` elements which contain any other additional attributes (e.g. `fetchpriority`, `crossorigin` or `data-do-not-generate-a-link-header`) will not be used to generate `Link` headers in order to prevent accidentally losing any custom prioritization logic that would otherwise be dropped as an Early Hint.

This allows you to directly create Early Hints as you're authoring your document, without needing to flick between your HTML and `_headers` file.

```html
---
filename: /blog/index.html
---

<html>
  <head>
    <link rel="preload" href="/style.css" as="stylesheet" />
    <link rel="stylesheet" href="/style.css" />
  </head>
</html>
```

> ### Disabling automatic `Link` header generation Automatic `Link` header

Reviewer: ```suggestion
### Disable automatic `Link` header generation Automatic `Link` header
```

> Automatic `Link` header generation should not have any negative performance impact on your website. If
> you find yourself needing to disable this feature, please get in touch with us!

Reviewer: ```suggestion
Automatic `Link` header generation should not have any negative performance impact on your website. If you find yourself needing to disable this feature, ____
```
How are they contacting us? Cloudflare account team? Support? Discord?

> You can remove any automatically generated `Link` headers by adding the following to your `_headers` file:

Reviewer: ```suggestion
Remove any automatically generated `Link` headers by adding the following to your `_headers` file:
```

Author: Same here.
